File: 684. Redundant Connection.py
class UF:
    def __init__(self, n):
        self.ids = [i for i in range(n)]
        
    def root(self, a):
        while a != self.ids[a]:
            self.ids[a] = self.ids[self.ids[a]]
            a = self.ids[a]
        return a
            
    def find(self, a, b):
        return self.root(a) == self.root(b)
    
    def union(self, a, b):
        ra = self.root(a)
        rb = self.root(b)
        self.ids[ra] = rb

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        uf = UF(len(edges)+1)
        for n1, n2 in edges:
            if uf.find(n1,n2):
                return [n1,n2]
            uf.union(n1, n2)

File: 684. Redundant Connection.java
class UF {
    int[] ids;
    public UF(int size) {
        ids = new int[size+1];
        for (int i=0;i<size+1;i++) {
            ids[i] = i;
        }
    }
    public int root(int A) {
        while (ids[A] != A) {
            ids[A] = ids[ids[A]];
            A = ids[A];
        }
        return A;
    }
    public boolean find(int A, int B) {
        return root(A) == root(B);
    }
    public void union(int A, int B) {
        int rA = root(A);
        int rB = root(B);
        ids[rA] = rB;
    }
}
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        UF uf = new UF(edges.length);
        int[] saved = null;
        for (int[] pair : edges) {
            if (uf.find(pair[0], pair[1])) {
                saved = pair;
                break;
            }
            uf.union(pair[0], pair[1]);
        }
        return saved;
    }
}